### PR TITLE
More work on APIs, demos and tests

### PR DESCRIPTION
**Pull request by cancerberoSgx**
I'm making a new PR so I can run the CI and make sure I don't screw it up. I would like to start thinking on merging this. I still need to solve some issues / complete the virtual command feature but I wanted to have your feedback now about. Besides that I think it's almost ready. See [changes](https://github.com/KnicKnic/WASM-ImageMagick/pull/28/files#diff-4ac32a78649ca5bdd8e0ba38b7006a1eR42)

### CODE REVIEW

File: tests/bundles/esm-test-cdn.html [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+  }
+  </style>
+
+  <h1>Testing /dist/bundles/index.esm.js import directly from html page </h1>
+
+  <p>You can import the library in modern browsers importing it as a JavaScript standard module: </p>
+
+  <pre>
+&lt;script type="module">    
+window.WASMImageMagickWorkerPath='https://unpkg.com/wasm-imagemagick@1.2.1/dist/magick.js'
```
[the remark this answers is not in the document]
Comment (KnicKnic) [reply tests/bundles/esm-test-cdn.html]: Doh' I wish I actually sent these comments when I wrote them.... Never mind.

**Comment by cancerberoSgx** [comment]
@KnicKnic I'm ready to push this into ML. Since it has lots of new features / refactors I would like you to verify it before pushing. See changes: https://github.com/KnicKnic/WASM-ImageMagick/pull/28/files#diff-4ac32a78649ca5bdd8e0ba38b7006a1eR42

File: webworker.js [hunk 2]
```diff
@@ -1,83 +1,114 @@
...
+if(notInAWebWorker){
+  throw new Error('magick.js must be loaded as a web worker, i.e. new Worker(pathToMagickJs)')
+}
+
 const stdout = []
 const stderr = []
 let exitCode = 0
 
+function getMagickJsUrl(path, prefix) {
+  debugger
```
Comment (KnicKnic) [inline webworker.js]: Did you want to leave this debugger statement in?

File: samples/interactive-execute-context/src/executeTemplate.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  singleImagePrefix?: string,
+  allImages?: string,
+  imageWidth?: string
+  imageHeight?: string
+  uniqueName?: string
+}
+export const defaultPlaceholders: CommandContextPlaceholders = {
+  singleImagePrefix: '$$IMAGE_',
+  allImages: '$$ALLIMAGES',
+  imageWidth: '$$IMAGE_WIDTH',
```
Comment (KnicKnic) [inline samples/interactive-execute-context/src/executeTemplate.ts]: I think I missed your reply on why we are adding our own templating for things like width when it should be covered in imagemagick? http://www.imagemagick.org/script/escape.php (%w)

File: src/magickApi.ts [hunk 4]
```diff
@@ -49,52 +183,99 @@ export interface CallResult {
...
+  })
+
   magickWorker.postMessage(request)
   magickWorkerPromisesKey++
-  return promise as Promise<CallResult>
+
+  return promise
 }
 
-function CreatePromiseEvent() {
```
Comment (KnicKnic) [inline src/magickApi.ts]: Cannot lose this function, it is a breaking change for my existing code.

**Comment by KnicKnic** [comment]
@cancerberoSgx so I did a few edits, but there are still some problems due to usage of require
The following line has an error in the produced magickApi.js 
`const Minimatch = require('minimatch');`
error
`Uncaught ReferenceError: require is not defined`
Repro url https://knicknic.github.io/imagemagick/rotate/index.1.html

**Comment by cancerberoSgx** [comment]
An update on this, I resume seeing the state of this branch but too much time have passed and I keep track of some features here. Although it builds and test correctly I'm afraid merging with main will have bad consequences since some features implemented here were also implemented  in master. Simply had too many features inside and too much time passed. I'm identifying features that could add value, and t I'm doing is identifying things that will add value, and for for each work on individual branches. The first one is adding support for text and fonts basically adding emscripten port for freetype. Next is pull a couple of features for  execute command syntax to support script-like strings little shell-like features so it's similar to scripts and easy to port them as a command. 
I'm not seeng right now much more relevant features but I need to revisit,
@KnicKnic   please don't delete the branch since I will be pulling code from there to individual PRs.

**Pull request closed without merging**